pyswarms/single/general_optimizer.py

```diff
@@ -0,0 +1,286 @@
+                parameter k must be included
+                * k : int
+                    number of neighbors to be considered. Must be a
+                    positive integer less than :code:`n_particles`
+                if used with the :code:`Ring` topology the additional
+                parameters k and p must be included
+                * p: int {1,2}
+                    the Minkowski p-norm to use. 1 is the
+                    sum-of-absolute values (or L1 distance) while 2 is
+                    the Euclidean (or L2) distance.
+        topology : pyswarms.backend.topology.Topology
+            a :code:`Topology` object that defines the topology to use
+            in the optimization process. The currently available topologies
+            are:
+                * Star
+                    All particles are connected
+                * Ring (static and dynamic)
+                    Particles are connected to the k nearest neighbours
+                * Pyramid (static and dynamic)
+                    Particles are connected in N-dimensional simplices
+                * Random (static and dynamic)
+                    Particles are connected to k random particles
+                Static variants of the topologies remain with the same neighbours
+                over the course of the optimization. Dynamic variants calculate
+                new neighbours every time step.
+        bounds : tuple of :code:`np.ndarray` (default is :code:`None`)
+            a tuple of size 2 where the first entry is the minimum bound
+            while the second entry is the maximum bound. Each array must
+            be of shape :code:`(dimensions,)`.
+        velocity_clamp : tuple (default is :code:`None`)
+            a tuple of size 2 where the first entry is the minimum velocity
+            and the second entry is the maximum velocity. It
+            sets the limits for velocity clamping.
+        center : list (default is :code:`None`)
+            an array of size :code:`dimensions`
+        ftol : float
+            relative error in objective_func(best_pos) acceptable for
+            convergence
+        """
+        super(GeneralOptimizerPSO, self).__init__(
+            n_particles,
+            dimensions=dimensions,
+            options=options,
+            bounds=bounds,
+            velocity_clamp=velocity_clamp,
+            center=center,
+            ftol=ftol,
+            init_pos=init_pos,
+        )
+
+        # Initialize logger
+        self.logger = logging.getLogger(__name__)
+        # Invoke assertions
+        self.assertions()
+        # Initialize the resettable attributes
+        self.reset()
+        # Initialize the topology and check for type
+        if not isinstance(topology, Topology):
+            raise TypeError("Parameter `topology` must be a Topology object")
+        else:
+            self.top = topology
+
+        # Case for the Ring topology
+        if isinstance(topology, Ring):
```

Comment:
Hmmm, in v.0.4.0, let’s make an errors module so that our assert statements and checks are abstracted. 

Better make VonNeumann explicit here just to be sure